api/src/main/java/org/pmiops/workbench/institution/InstitutionServiceImpl.java

```diff
@@ -334,27 +356,41 @@ private void populateAuxTables(
 
   // note that this replaces all email domains for this institution with the passed-in domains
   private void setInstitutionEmailDomains(
-      final Institution modelInstitution, final DbInstitution dbInstitution) {
+      final Institution modelInstitution,
+      final DbInstitution dbInstitution,
+      final List<DbAccessTier> dbAccessTiers) {
     institutionEmailDomainDao.deleteByInstitution(dbInstitution);
-    institutionEmailDomainMapper
-        .modelToDb(modelInstitution, dbInstitution)
-        .forEach(institutionEmailDomainDao::save);
+    for (DbAccessTier dbAccessTier : dbAccessTiers) {
+      institutionEmailDomainMapper
+          .modelToDb(modelInstitution, dbInstitution, dbAccessTier)
+          .forEach(institutionEmailDomainDao::save);
+    }
   }
 
   // note that this replaces all email addresses for this institution with the passed-in addresses
   private void setInstitutionEmailAddresses(
-      final Institution modelInstitution, final DbInstitution dbInstitution) {
+      final Institution modelInstitution,
+      final DbInstitution dbInstitution,
+      final List<DbAccessTier> dbAccessTiers) {
     institutionEmailAddressDao.deleteByInstitution(dbInstitution);
-    institutionEmailAddressMapper
-        .modelToDb(modelInstitution, dbInstitution)
-        .forEach(institutionEmailAddressDao::save);
+    for (DbAccessTier dbAccessTier : dbAccessTiers) {
+      institutionEmailAddressMapper
+          .modelToDb(modelInstitution, dbInstitution, dbAccessTier)
+          .forEach(institutionEmailAddressDao::save);
+    }
   }
 
   // note that this replaces all requirements for this institution with the passed-in requirements
   private void setInstitutionTierRequirement(
-      final Institution modelInstitution, final DbInstitution dbInstitution) {
-    List<DbAccessTier> dbAccessTiers = accessTierDao.findAll();
+      final Institution modelInstitution,
+      final DbInstitution dbInstitution,
+      final List<DbAccessTier> dbAccessTiers) {
     institutionTierRequirementDao.deleteByInstitution(dbInstitution);
+    // Make sure the delete success.
+    if (!institutionTierRequirementDao.getByInstitution(dbInstitution).isEmpty()) {
```

Comment:
Why is this needed? Would this be better as unit test instead? Is there a reason to think that a `delete` method won't actually delete the row?

Reply:
I added a per_tier_per_institution constraints in tier_requirement table. And seem delete then insert does not work because the constraints is not removed from DB yet. Seems deleting the actual column and deleting the constraints is not handled in the same transaction. I tried adding retry in the insert below still not working. 
I figured I can make a get call, somehow hibernaate can flush commit, and I am able to make insert call successfully.  I am not 100% sure why it works in this way..